allennlp/common/from_params.py

```diff
@@ -128,23 +128,46 @@ def create_kwargs(
     signature = inspect.signature(constructor)
     kwargs: Dict[str, Any] = {}
 
+    parameters = dict(signature.parameters)
+
+    # First we check for the presence of a **kwargs parameter.  If we find one, we look in the
+    # superclass constructor, to see if there are arguments there that we should try to also
+    # constructor.
+    has_kwargs = False
+    for param in parameters.values():
+        if param.kind == param.VAR_KEYWORD:
+            has_kwargs = True
+    if has_kwargs:
+        # "mro" is "method resolution order".  The first one is the current class, the next is the
+        # first superclass, and so on.  Taking the first superclass should work in all cases that
+        # we're looking for here.
+        superclass = cls.mro()[1]
+        superclass_signature = inspect.signature(superclass.__init__)  # type: ignore
+        for param_name, param in superclass_signature.parameters.items():
+            if param_name == "self":
+                continue
+            parameters[param_name] = param
```

Comment:
I think overwriting is not a good idea here. The correct way is to infer only those parameters that couldn't be inferred at the first step.
An example:
```python
class Some(DatasetReader):
    def __init__(self, some_param: str, **kwargs):
        super().__init__(**kwargs)
        self._some_param = some_param

class B(A):
    def __init__(self, some_param: str = 'default_value', **kwargs):
        super().__init__(some_param=some_param, **kwargs)
```

Reply:
PR #3653 tries to fix this issue.